airflow/contrib/hooks/gcs_hook.py

```diff
@@ -564,6 +564,167 @@ def compose(self, bucket_name, source_objects, destination_object):
+                    destination_object=dst_object,
+                )
+            self.log.info("Blobs copied.")
+        # Delete redundant files
+        if to_delete_blobs:
+            if delete_extra_files:
+                # TODO: Add batch. I tried to do it, but the Google library is not stable at the moment.
+                for blob in to_delete_blobs:
+                    self.delete(blob.bucket.name, blob.name)
+                self.log.info("Blobs deleted.")
+            else:
+                self.log.info("Skipped blobs deleting")
+
+        # Overwrite files that are different
+        if to_rewrite_blobs:
+            if allow_overwrite:
+                for blob in to_rewrite_blobs:
+                    dst_object = (
+                        path.join(destination_object, blob.name[source_object_prefix_len:])
+                        if destination_object
+                        else blob.name[source_object_prefix_len:]
+                    )
+                    self.rewrite(
+                        source_bucket=source_bucket_obj.name,
+                        source_object=blob.name,
+                        destination_bucket=destination_bucket_obj.name,
+                        destination_object=dst_object,
+                    )
+                self.log.info("Blob rewritten.")
+            else:
+                self.log.info("Skipped blobs overwriting")
+        self.log.info("Synchronization finished")
+
+    @staticmethod
+    def _prepare_sync_plan(
+        source_bucket: storage.Bucket,
+        destination_bucket: storage.Bucket,
+        source_object: Optional[str],
+        destination_object: Optional[str],
+        recursive: bool,
+    ) -> Tuple[Set[storage.Blob], Set[storage.Blob], Set[storage.Blob]]:
+        # Calculate the number of characters that remove from the name, because they contain information
+        # about the parent's path
+        source_object_prefix_len = len(source_object) if source_object else 0
+        destination_object_prefix_len = len(destination_object) if destination_object else 0
+        delimiter = "/" if not recursive else None
+        # Fetch blobs list
+        source_blobs = list(source_bucket.list_blobs(prefix=source_object, delimiter=delimiter))
+        destination_blobs = list(
+            destination_bucket.list_blobs(prefix=destination_object, delimiter=delimiter))
+        # Create sets with names without parent object name
+        source_names = {a.name[source_object_prefix_len:] for a in source_blobs}
+        destination_names = {a.name[destination_object_prefix_len:] for a in destination_blobs}
+        # Create indexes that allow you to identify blobs based on their name
+        source_names_index = {a.name[source_object_prefix_len:]: a for a in source_blobs}
+        destination_names_index = {a.name[destination_object_prefix_len:]: a for a in destination_blobs}
+        # Determine objects to copy and delete
```

Comment:
```suggestion
        # Determine objects to copy and delete
        source_names = set(source_names_index.keys())
        destination_names = set(destination_names_index.keys())
```
Then lines 706 and 707 are redundant and we iterate only once.

Reply:
I updated this PR.